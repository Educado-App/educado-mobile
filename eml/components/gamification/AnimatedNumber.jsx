import React, { useRef, useEffect, useState } from 'react';
import { View } from 'react-native';
import Animated, { EasingNode } from 'react-native-reanimated';
import Text from '../general/Text';
import PropTypes from 'prop-types';	

//inspo from https://www.npmjs.com/package/react-native-animated-numbers?activeTab=code

/* This is basically a package, copied and modified */

// makes an array of numbers from 0 to 9
const NUMBERS = Array(10).fill().map((_, i) => i);

// custom hook to get the previous number
const usePrevious = (value) => {
	const ref = useRef();
	useEffect(() => {
		ref.current = value;
	});

	if (typeof ref.current === 'undefined') {
		return 0;
	}

	return ref.current;
};

const AnimatedNumber = ({
	animateToNumber,
	fontStyle,
	animationDuration,
	easing,
}) => {
	// the previous number
	const prevNumber = usePrevious(animateToNumber);

	// takes the number and converts it to a string
	const animateToNumberString = String(Math.abs(animateToNumber));

	// takes the previous number and converts it to a string
	const prevNumberString = String(Math.abs(prevNumber));

	// splits the string into an array of numbers, thus 100 is [1,0,0]
	const animateToNumbersArr = Array.from(animateToNumberString, Number);

	const prevNumberersArr = Array.from(prevNumberString, Number);

	// the height of the text element
	const [numberHeight, setNumberHeight] = useState(0);

	// creates an array of animated values, one for each number e.g. 50 is 2 animated values
	const animations = animateToNumbersArr.map((__, index) => {
		// if the number is not a number (there is no prevNumber) then return 0
		if (typeof prevNumberersArr[index] !== 'number') {
			return new Animated.Value(0);
		}

		// the height of the animation is the height of the text element * the number 
		//e.g. if the text element is 25px high and the number is 2 then the height of the animation is 50px so the animation moves 50px up
		const animationHeight = -1 * (numberHeight * prevNumberersArr[index]);
		return new Animated.Value(animationHeight);
	});

	// when the text element is created, the height is set
	const setButtonLayout = (e) => {
		setNumberHeight(e.nativeEvent.layout.height);
	};

	// when the number changes, the animation is triggered
	useEffect(() => {
		// maps through the animations array and sets the animation e.g. 50 is 2 animations
		animations.map((animation, index) => {
			// if the number is not a number then return
			if (typeof animateToNumbersArr[index] !== 'number') {
				return;
			}

			Animated.timing(animation, {
				// the animation moves the number up by the height of the text element * the number
				toValue: -1 * (numberHeight * animateToNumbersArr[index]),
				duration: animationDuration || 1400,
				useNativeDriver: true,
				easing: easing || EasingNode.elastic(1.2),
			}).start();
		});
	}, [animateToNumber, numberHeight]);

	// the translateY value for each number, how much it should move in the y axis
	const getTranslateY = (index) => {
		return animations[index];
	};

	return (
		<>
			{numberHeight !== 0 && (
			// creates a row 
				<View style={{ flexDirection: 'row' }}>
					{/* if negative number then - is viewed */}
					{animateToNumber < 0 && (
						<Text className={`${fontStyle}`} style={{ height: numberHeight }}>{'-'}</Text>
					)}
					{animateToNumbersArr.map((n, index) => {
						{/* if the one of indexes in the array is string then return it as is */}
						if (typeof n === 'string') {
							return (
								<Text key={index} className={`${fontStyle}`} style={{ height: numberHeight }}>
									{String(n)}
								</Text>
							);
						}

						// else return the animated.view
						return (
						// the view for the numbers where all besides the current number are hidden
							<View
								key={index}
								style={{ height: numberHeight, overflow: 'hidden' }}
							>
								{/* animate to the next number based */}
								<Animated.View
									style={[
										{
											transform: [
												{
													translateY: getTranslateY(index),
												},
											],
										},
									]}
								>
									{/* prints all the numbers from 0-9 but hides the unused numbers */}
									{NUMBERS.map((number, i) => (
										<View key={i} className="w-fit">
											<Text className={`${fontStyle}`} style={{ height: numberHeight }}>
												{String(number)}
											</Text>
										</View>
									))}
								</Animated.View>
							</View>
						);
					})}
				</View>
			)}
			<Text 
				className={`${fontStyle}`}
				style={{ position: 'absolute', top: -999999 }}
				onLayout={setButtonLayout}
			>
				{String(0)}
			</Text>
		</>
	);
};

AnimatedNumber.propTypes = {
	animateToNumber: PropTypes.number,
	fontStyle: PropTypes.string,
	animationDuration: PropTypes.number,
	easing: PropTypes.func,
};

export default AnimatedNumber;