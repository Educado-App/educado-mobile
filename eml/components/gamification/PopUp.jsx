import React, { useState, useEffect } from 'react';
import { View, Animated, Easing } from 'react-native';
import Text from '../general/Text';

export function PopUp({ randomPhrase, xpAmount, isCorrectAnswer }) {
  const animatedValue = new Animated.Value(0);

  const startAnimation = () => {
    Animated.timing(animatedValue, {
      toValue: 1,
      duration: 250, // Adjust the duration as needed
      easing: Easing.ease, // Use a suitable easing function
      useNativeDriver: false, // Set to true if possible
    }).start(() => {
      setTimeout(() => {
        hidePopup();
      }, 5000); // Hide after 5 seconds
    });
  };

  const hidePopup = () => {
    Animated.timing(animatedValue, {
      toValue: 0,
      duration: 400, // Adjust the duration as needed
      easing: Easing.ease, // Use a suitable easing function
      useNativeDriver: false, // Set to true if possible
    }).start();
  };

  useEffect(() => {
    startAnimation(); // Call the animation when the component is rendered
  }, []);

  return (
    <Animated.View
      style={{
        opacity: animatedValue, // Apply the animated value to opacity
        transform: [
          {
            translateY: animatedValue.interpolate({
              inputRange: [0, 1],
              outputRange: [100, 0], // Adjust the values for the desired effect
            }),
          },
        ],
      }}
      className="bg-bgPrimary absolute bottom-0 p-4 w-screen h-[12.5%] z-20 rounded-large shadow-md shadow-projectBlack flex-row justify-between"
      testID={'PopUp'}
    >
      <View className={`${isCorrectAnswer === true ? 'w-10/12' : 'w-full'}`}>
        <Text className={`font-sans-bold text-sm ${isCorrectAnswer === true ? 'text-correctAnswer' : 'text-wrongAnswer'}`}>{randomPhrase}</Text>
      </View>

      {isCorrectAnswer === true ?
        <Text className='font-sans-bold text-sm text-correctAnswer'>{xpAmount}xp</Text>
        : null
      }

    </Animated.View>
  );
}

export function XpPopUp({ randomPhrase, xpAmount, isCorrectAnswer }) {
  const animatedValue = new Animated.Value(0);
  const opacityValue = new Animated.Value(1);
  const timer = 3000;
  const randomVariation = Math.random() * 200; // Adjust the range as needed
  
  const customEasing = (value) => {
    // Adjust these parameters to control the easing effect
    const power = 2; // Determines the rate of change. Higher values result in a faster initial acceleration.
    const factor = 2; // A factor to adjust the curve
    const delay = 5; // Delay factor for slower start

    return (
      Math.pow(delay + (1 - delay) * value, power) /
      (Math.pow(delay + (1 - delay) * value, power) +
        Math.pow(factor - factor * (delay + (1 - delay) * value), power))
    );
  };

  const startAnimation = () => {
    Animated.parallel([
      Animated.timing(animatedValue, {
        toValue: 8,
        duration: timer,
        easing: (value) => customEasing(value), // Use the custom easing function here
        useNativeDriver: false,
      }),
      Animated.timing(opacityValue, {
        toValue: 0,
        duration: timer,
        easing: (value) => customEasing(value), // Use linear easing for opacity to maintain consistency
        useNativeDriver: false,
      }),
    ]).start(() => {
      // Hide after 2 seconds (2000 milliseconds)
      setTimeout(() => {
        // Add your code to handle the completion of the animation if needed
      }, timer);
    });
  };

  useEffect(() => {
    startAnimation(); // Call the animation when the component is rendered
  }, []);

  return (
    <Animated.View
      style={{
        opacity: animatedValue,
        transform: [
          {
            translateY: animatedValue.interpolate({
              inputRange: [0, 1],
              outputRange: [100, 0], // Adjust the values for the desired effect
            }),
          },
          {
            translateX: randomVariation,
          },
        ],
      }}
      className="absolute bottom-0 w-screen h-[12.5%] z-20 shadow-md shadow-projectBlack"
      testID={'XpPopUp'}
    >
      {isCorrectAnswer === true ? (
        <Animated.Text
          style={{
            opacity: opacityValue,
          }}
          className="font-sans-bold text-lg text-correctAnswer text-center"
        >
          {xpAmount}xp
        </Animated.Text>
      ) : null}
    </Animated.View>
  );
}
