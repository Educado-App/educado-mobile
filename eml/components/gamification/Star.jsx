import React from 'react';
import { Icon } from '@rneui/themed';

export default function star() {
	return (
		<Icon
			style={{
				shadowOpacity: 0.8,
				shadowRadius: 2,
				shadowOffset: { width: 0, height: 0 },
				shadowColor: '#ffd633'
			}}
			size={30}
			name="star"
			type="material-community"
			color="#ffd633"
		/>
	);
}
