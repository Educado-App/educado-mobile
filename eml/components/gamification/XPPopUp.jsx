import React, { useState, useEffect } from 'react';
import { View, Animated, Easing } from 'react-native';
import Text from '../general/Text';

export default function PopUp({ randomPhrase, xpAmount, isCorrectAnswer }) {
  const animatedValue = new Animated.Value(0);
  const opacityValue = new Animated.Value(1);
  const timer = 3000;
  const randomVariation = Math.random() * 200; // Adjust the range as needed
  
  const customEasing = (value) => {
    // Adjust these parameters to control the easing effect
    const power = 2; // Determines the rate of change. Higher values result in a faster initial acceleration.
    const factor = 2; // A factor to adjust the curve
    const delay = 5; // Delay factor for slower start

    return (
      Math.pow(delay + (1 - delay) * value, power) /
      (Math.pow(delay + (1 - delay) * value, power) +
        Math.pow(factor - factor * (delay + (1 - delay) * value), power))
    );
  };

  const startAnimation = () => {
    Animated.parallel([
      Animated.timing(animatedValue, {
        toValue: 8,
        duration: timer,
        easing: (value) => customEasing(value), // Use the custom easing function here
        useNativeDriver: false,
      }),
      Animated.timing(opacityValue, {
        toValue: 0,
        duration: timer,
        easing: (value) => customEasing(value), // Use linear easing for opacity to maintain consistency
        useNativeDriver: false,
      }),
    ]).start(() => {
      // Hide after 2 seconds (2000 milliseconds)
      setTimeout(() => {
        // Add your code to handle the completion of the animation if needed
      }, timer);
    });
  };

  useEffect(() => {
    startAnimation(); // Call the animation when the component is rendered
  }, []);

  return (
    <Animated.View
      style={{
        opacity: animatedValue,
        transform: [
          {
            translateY: animatedValue.interpolate({
              inputRange: [0, 1],
              outputRange: [100, 0], // Adjust the values for the desired effect
            }),
          },
          {
            translateX: randomVariation,
          },
        ],
      }}
      className="absolute bottom-0 w-screen h-[12.5%] z-20 shadow-md shadow-projectBlack"
    >
      {isCorrectAnswer === true ? (
        <Animated.Text
          style={{
            opacity: opacityValue,
          }}
          className="font-sans-bold text-lg text-correctAnswer text-center"
        >
          {xpAmount}xp
        </Animated.Text>
      ) : null}
    </Animated.View>
  );
}
