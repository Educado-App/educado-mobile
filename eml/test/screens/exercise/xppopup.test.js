import React from 'react';
import renderer from 'react-test-renderer';
import { XpPopUp } from '../../../components/gamification/PopUp';

let XpPopUpComponent;

beforeEach(() => {
  XpPopUpComponent = renderer.create(
    <XpPopUp randomPhrase="Test Phrase" xpAmount={100} isCorrectAnswer={true} />
  );
});

describe('XpPopUp', () => {
  it('renders XpPopUp correctly for correct answer', () => {
    expect(XpPopUpComponent.toJSON()).toMatchSnapshot();
  });

  it('renders XpPopUp correctly for incorrect answer', () => {
    const IncorrectXpPopUpComponent = renderer.create(
      <XpPopUp randomPhrase="Test Phrase" xpAmount={0} isCorrectAnswer={false} />
    );
    expect(IncorrectXpPopUpComponent.toJSON()).toMatchSnapshot();
  });
});
