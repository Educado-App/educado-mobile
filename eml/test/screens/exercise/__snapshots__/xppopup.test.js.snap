// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XpPopUp renders XpPopUp correctly for correct answer 1`] = `
<View
  collapsable={false}
  style={
    {
      "bottom": 0,
      "height": "12.5%",
      "opacity": 0,
      "position": "absolute",
      "shadowColor": "#383838",
      "shadowOffset": {
        "height": 6,
        "width": 0,
      },
      "shadowOpacity": 1,
      "shadowRadius": 10,
      "transform": [
        {
          "translateY": 100,
        },
        {
          "translateX": 71.42850506722938,
        },
      ],
      "width": 750,
      "zIndex": 20,
    }
  }
  testID="XpPopUp"
>
  <Text
    collapsable={false}
    style={
      {
        "color": "#00897B",
        "fontFamily": "Montserrat-Bold",
        "fontSize": 18,
        "lineHeight": 28,
        "opacity": 1,
        "textAlign": "center",
      }
    }
  >
    100
    xp
  </Text>
</View>
`;

exports[`XpPopUp renders XpPopUp correctly for incorrect answer 1`] = `
<View
  collapsable={false}
  style={
    {
      "bottom": 0,
      "height": "12.5%",
      "opacity": 0,
      "position": "absolute",
      "shadowColor": "#383838",
      "shadowOffset": {
        "height": 6,
        "width": 0,
      },
      "shadowOpacity": 1,
      "shadowRadius": 10,
      "transform": [
        {
          "translateY": 100,
        },
        {
          "translateX": 91.97157056466692,
        },
      ],
      "width": 750,
      "zIndex": 20,
    }
  }
  testID="XpPopUp"
/>
`;
